### ISSUE TITLE
Huge CPU load, NCPU has no effect?

### ISSUE THREAD

**Issue by amisk** [mhardcastle/ddf-pipeline#172]
Hi,

I am running the ddf pipeline. I have 48 cores. The 1 minute CPU load is always around 700 when kMs.py is running. I think that is too much and might create too much overhead. Dstat shows around 20% usr, and 80% sys usage.

I then tried using NCPU = 16, but according to htop, kms.py still uses 48 cores. Is there a different parameter specifically for kms.py?

Thx.

**Comment by mhardcastle** [comment]
Hm, that should not happen (especially 80% sys). Have you taken the steps described in the installation guide (specifically on hugepage defragmentation)?

**Comment by amisk** [comment]
Yes, I have the lines 

```
mount -o remount -o size=256g /dev/shm
echo never > /sys/kernel/mm/transparent_hugepage/defrag

```
in /etc/rc.local.

Is it possible that this has no effect in the latest ubuntu versions?

**Comment by cyriltasse** [comment]
If you `OpenBLAS` and not `ATLAS` you should tell him to only use one CPU per process, setting this

```
export OPENBLAS_NUM_THREADS=1
export OPENBLAS_MAX_THREADS=1
```

**Comment by cyriltasse** [comment]
By the way this is a `kMS` issue, not `ddf-pipeline`

**Comment by amisk** [comment]
I appear to be using ATLAS, libatlas-base-dev and libatlas-base3 is installed, libopenblas is not installed.

**Comment by cyriltasse** [comment]
ok - might not be that... if you use `--NCPU=1` when you `htop` during the solve, what do you see?

**Comment by amisk** [comment]
Ok, so if I start kms with --NCPU=1 it only uses one core. If I start the pipeline, it starts kms with --NCPU=48, altough I have 
```
[machine]
NCPU = 16

```

In the config file. Is it important where this is in the config file?

**Comment by cyriltasse** [comment]
Hoooo - indeed a `ddf-pipeline` issue then, apologies. But since it's at the script level, it should be trivial to solve then...

**Comment by mhardcastle** [comment]
In the version of the pipeline I'm looking at there are separate NCPU_killms and NCPU_DDF options.

**Comment by amisk** [comment]
It works with NCPU_killms. Thanks.

But, I've been thinking, since I do not know much about cpu/thread architectures.
Even if now 16 cpus are used, is it not still possible that those 16 cpus are still bombarded with too many threads, so that it will still be much too slow?

**Comment by amisk** [comment]
Ah, here is what I mean by that. Two screenshots, one showing htop, one showing dstat. Apologies for the format, at the moment I can only log in via phone.
![IMG_20190425_150706](https://user-images.githubusercontent.com/19927617/56738205-692a0080-676c-11e9-9d99-0ca0858f5cb8.jpg)
![IMG_20190425_150742](https://user-images.githubusercontent.com/19927617/56738206-692a0080-676c-11e9-9b35-73faea8b6d5c.jpg)

In htop, about half of the cpu bar of the 16 cpus are red, which I think is system time. Also in dstat, sys is at 26 or so, while usr has little time.

**Comment by amisk** [comment]
I will try now to install OpenBlas and recompile killms to use this instead of atlas.
Would that break something else?

**Comment by mhardcastle** [comment]
Isn't the issue what numpy is built with?

**Comment by amisk** [comment]
So, I just installed libopenblas-dev and libopenblase-base. I only recompiled DDfacet and Killms. I really did nothing else.
Now it appears to use OpenBlas, because I used the export line to limit the number of threads, and now it appears to run very smoothly. The load is now at around 50 when using all 48 cores, and dstat shows about 75/25 percent for usr/sys time.
Lets see if it will now be faster than the 23 days I am already for it to finish :)
